src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs

```diff
@@ -0,0 +1,701 @@
+        {
+            var readScores = MemoryMarshal.Cast<long, float>(batchTermIds)[..batchResults.Length];
+
+            // We have to initialize the score buffer with a positive number to ensure that multiplication (document-boosting) is taken into account when BM25 relevance returns 0 (for example, with AllEntriesMatch).
+            readScores.Fill(Bm25Relevance.InitialScoreValue);
+
+            // We perform the scoring process. 
+            match._inner.Score(batchResults, readScores, 1f);
+
+            // If we need to do documents boosting then we need to modify the based on documents stored score. 
+            if (match._searcher.DocumentsAreBoosted)
+            {
+                // We get the boosting tree and go to check every document. 
+                BoostDocuments(match, batchResults, readScores);
+            }
+
+            // Note! readScores & indexes are aliased and same as batchTermIds
+            var indexes = MemoryMarshal.Cast<long, int>(batchTermIds)[..(batchTermIds.Length)];
+            for (int i = 0; i < batchTermIds.Length; i++)
+            {
+                batchTerms[i] = new UnmanagedSpan(readScores[i]);
+                indexes[i] = i;
+            }
+
+            EntryComparerHelper.IndirectSort<EntryComparerByScore, TComparer2, TComparer3>(ref match, indexes, batchTerms, new(), comparer2, comparer3);
+
+            for (int i = 0; i < indexes.Length; i++)
+            {
+                match._results.Add(batchResults[indexes[i]]);
+            }
+        }
+
+        private static void BoostDocuments(SortingMultiMatch<TInner> match, Span<long> batchResults, Span<float> readScores)
+        {
+            var tree = match._searcher.GetDocumentBoostTree();
+            if (tree is {NumberOfEntries: > 0})
+            {
+                // We are going to read from the boosting tree all the boosting values and apply that to the scores array.
+                ref var scoresRef = ref MemoryMarshal.GetReference(readScores);
+                ref var matchesRef = ref MemoryMarshal.GetReference(batchResults);
+                for (int idx = 0; idx < batchResults.Length; idx++)
+                {
+                    var ptr = (float*)tree.ReadPtr(Unsafe.Add(ref matchesRef, idx), out var _);
+                    if (ptr == null)
+                        continue;
+
+                    ref var scoresIdx = ref Unsafe.Add(ref scoresRef, idx);
+                    scoresIdx *= *ptr;
+                }
+            }
+        }
+
+        public int Compare(UnmanagedSpan x, UnmanagedSpan y)
+        {
+            // Note, for scores, we go *descending* by default!
+            return y.Double.CompareTo(x.Double);
+        }
+
+        public int Compare(int x, int y)
+        {
+            throw new NotSupportedException(ScoreComparerAsInnerExceptionMessage);
+        }
+    }
+
+    private struct CompactKeyComparer : IComparer<UnmanagedSpan>
```

Comment:
This is duplicate without any reason from SortingMatch, right?

Reply:
@maciejaszyk ping

Reply:
Yes, 
I created a ticket to take a look into all code duplications.
https://github.com/ravendb/ravendb/pull/16679#discussion_r1238223175